### ISSUE TITLE
2.0.17/2.0.18 regression: Undefined symbols `htons` and `ntohs`

### ISSUE THREAD

**Issue by Faless** [emscripten-core/emscripten#14012]
We've been experiencing this issue in Godot builds with emscripten `2.0.17+` (`2.0.16` works).

```
[100%] error: undefined symbol: htons (referenced by $inetPton6__deps: ['htons','$jstoi_q'], referenced by $writeSockaddr__deps: ['$Sockets','$inetPton4','$inetPton6'], referenced by __sys_accept4__deps: ['$getSocketFromFD','$writeSockaddr','$DNS','$SYSCALLS'], referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _htons may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: ntohs (referenced by $inetNtop6__deps: ['$inetNtop4','ntohs'], referenced by $readSockaddr__deps: ['$Sockets','$inetNtop4','$inetNtop6','ntohs'], referenced by $getSocketAddress__deps: ['$readSockaddr'], referenced by __sys_bind__deps: ['$getSocketFromFD','$getSocketAddress','$SYSCALLS'], referenced by top-level compiled C/C++ code)
warning: _ntohs may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
em++: error: '/media/Storage/emsdk/node/14.15.5_64bit/bin/node /media/Storage/emsdk/upstream/emscripten/src/compiler.js /tmp/tmptx_3wswa.txt' failed (1)
scons: *** [bin/godot.javascript.opt.debug.js] Error 1
scons: building terminated because of errors.
```

**Comment by kripken** [comment]
Does this still happen on 2.0.18?

My best guess is that this might be a regression from @sbc100 's REVERSE_DEPS and other deps_info changes. But I think those were before 2.0.17.

@Faless Bisection to find the specific bad change might speed this up, https://emscripten.org/docs/contributing/developers_guide.html#bisecting

**Comment by Faless** [comment]
I'll look into bisecting. I thought about that change being the source too,
but I can confirm that 2.0.16 works, while 2.0.17 does not.

**Comment by Faless** [comment]
And yes, Sorry, I also confirm 2.0.18 is failing

**Comment by sbc100** [comment]
So there is an entry in deps_info.py for this: https://github.com/emscripten-core/emscripten/blob/897d6fac93c4152feb03dfbef6eec1e21af39ae1/tools/deps_info.py#L179

At least if you program calls `accept` to should also pull in htons.

I wonder if `__sys_accept_4` is being called without the user code including a call to `accept`?   Do you know why/how your program depends on `__sys_accept_4`?

**Comment by sbc100** [comment]
Ah.. I guess your application could be calling `accept4` (rather than accept)?   However, if so, it would be interesting to know this worked before.

**Comment by sbc100** [comment]
By far the most likely candidate for the breakage is #13483.   However, that change landed in 2.0.15.. which is odd

**Comment by Faless** [comment]
Well, we do call accept:
https://github.com/godotengine/godot/blob/a6651de537f5729262d8c8beb678e886c86c188b/drivers/unix/net_socket_posix.cpp#L724

Could the problem be that we use it as `::accept` inside another
`Class::accept`?
Not sure how dependency parsing works, but I assume it's on the generated
symbols, not the source code.

**Comment by sbc100** [comment]
Yes, we run `llvm-nm` on all your input objects and libraries and loop for symbols like `accept` .. if we find them to be undefined in your input files that triggers the inclusion of (in this case) `htons`.

Could you include the full output of the link step wit h`EMCC_DEBUG=1` set?

**Comment by Faless** [comment]
Here's the log, it does seem to reference `__sys_accept4` and not `__sys_accept`, but I don't know why

[emcc.zip](https://github.com/emscripten-core/emscripten/files/6383366/emcc.zip)

**Comment by Faless** [comment]
Interestingly, `2.0.16` also reference `__sys_accept4`, but the build does work:

[emcc_2.0.16.zip](https://github.com/emscripten-core/emscripten/files/6383806/emcc_2.0.16.zip)

**Comment by sbc100** [comment]
Can you see if adding the line `'accept4': ['htons'],`  to `tools/deps_info.py` fixes the issue.    We already have the line `'accept': ['htons']` which means that programs that use `accept` already get `'htons'` included, but its possible your program is directly calling accept?

Also the output of the llvm-nm command that we see in the log would tell is all of the symbols that your program is pulling in:

```
shared:DEBUG: executed /media/Storage/emsdk/upstream/bin/llvm-nm /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/audio_driver_javascript.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/http_client_javascript.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/javascript_eval.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/javascript_main.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/os_javascript.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/api/javascript_tools_editor_plugin.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/javascript/javascript_runtime.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/main/libmain.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/modules/libmodules.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/platform/libplatform.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/drivers/libdrivers.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/scene/libscene.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/servers/libservers.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/core/libcore.javascript.opt.debug.bc /home/fales/Programming/workspaces/git/godot-3.2/modules/freetype/libfreetype_builtin.javascript.opt.debug.bc  
```

Can you see if accept or accept4 appear in the list of undefined symbols (symbols that start with `U`)?

**Comment by Faless** [comment]
This is so weird. I can see `U accept` in both logs.

[undef_2.0.16.zip](https://github.com/emscripten-core/emscripten/files/6385104/undef_2.0.16.zip)
[undef_2.0.17.zip](https://github.com/emscripten-core/emscripten/files/6385105/undef_2.0.17.zip)

Additionally, we also use `recvfrom` and `send`, which also add deps for `htons` and `ntohs`.
I tried adding `accept4` to the `deps_info.py` but I'm still getting the same error.

**Comment by sbc100** [comment]
In that case it looks like maybe we have a bug in `handle_reverse_deps` .. that function is supposed to take the output `nm` and then include the reverse deps as needed.   There have been some recent changes to way we run `nm` so it could be that.   Can you try to debug that function a little?  i.e. see if `accept` is showing up in the output of `building.llvm_nm_multiple` and if so why its not pulling in `htons` is it should?

**Comment by Faless** [comment]
The output of `building.llvm_nm_multiple` `undefs` is many times smaller in `2.0.17` then in `2.0.16`, and it does not include `accept` (while the one in `2.0.16` does).

**Comment by sbc100** [comment]
@juj did some refactoring there recently and might have some insight.  

Any more info you can glean from debugging would be very useful @Faless

**Comment by Faless** [comment]
Bisected:
19957ab141aff19af6bfbfbde9d25fcdf589ae8a is the first bad commit.

```
git bisect start
# good: [708a5a76766e0d7d710032b4a06c87cae0b74a76] 2.0.16 (#13770)
git bisect good 708a5a76766e0d7d710032b4a06c87cae0b74a76
# bad: [f039cfa13a3c7e2045eb96df28785e20694af402] Address review for layout(binding) support: more precise regex, and avoid ES6 arrow notation. (#13984)
git bisect bad f039cfa13a3c7e2045eb96df28785e20694af402
# bad: [26fbab35cfd2da224093bbf87b7212be7a3e8154] Deprecate EXTRA_EXPORTED_RUNTIME_METHODS (#13895)
git bisect bad 26fbab35cfd2da224093bbf87b7212be7a3e8154
# bad: [6f8dc887cd4693676adeab51af6b1cd6f5da096f] Simplify input file handling in emcc. NFC. (#13799)
git bisect bad 6f8dc887cd4693676adeab51af6b1cd6f5da096f
# good: [0aef720ece71c9950bd388f226c21a2f9be75d04] New uniform location handling (#13732)
git bisect good 0aef720ece71c9950bd388f226c21a2f9be75d04
# bad: [04b2a82ad62022c0bc65f714c1e55e5e9df948dd] Switch to the new LLVM pass manager (#13818)
git bisect bad 04b2a82ad62022c0bc65f714c1e55e5e9df948dd
# bad: [130e6e1cb38be58ebbbe7451d528b4a46aa73185] Fix stale comment (#13805)
git bisect bad 130e6e1cb38be58ebbbe7451d528b4a46aa73185
# bad: [19957ab141aff19af6bfbfbde9d25fcdf589ae8a] No system lib multiprocessing (#13493)
git bisect bad 19957ab141aff19af6bfbfbde9d25fcdf589ae8a
# good: [f1bf33c8e9bb3cb063a038b404eb21a1ccecac89] Fix lack of JS function mapping for symbols file generation for WASM=0 and WASM=2 with --emit-symbol-map (#13782)
git bisect good f1bf33c8e9bb3cb063a038b404eb21a1ccecac89
# first bad commit: [19957ab141aff19af6bfbfbde9d25fcdf589ae8a] No system lib multiprocessing (#13493)
```

**Comment by kripken** [comment]
Interesting, thanks @Faless 

@juj this looks like your multiprocessing changes broke this. Perhaps the nm part specifically? Worrying that it is affected in this way...

**Comment by Faless** [comment]
Btw, if you want to try the build:

```
git clone git@github.com:godotengine/godot.git --depth 1 --branch 3.x
cd godot
scons p=javascript target=release_debug
```

Optionally adding `-jNUM` to use more cores

**Comment by juj** [comment]
Apologies for the regression @Faless .

The issue here looks like is that Godot is generating static library archive files (`.a`) for link, but with the incorrect suffix `.bc` that represent LLVM bitcode files. The new llvm-nm logic is using the file suffix to determine the parsing code, and is then thrown off when it sees .bc files that are instead .a files.

Posted a fix in #14053 which simplifies the parsing so that it does not need to assume anything based on the suffix. 

Independently Godot may want to change the build system to use `.a` file suffixes for static libraries.

**Comment by Faless** [comment]
> The issue here looks like is that Godot is generating static library archive files (`.a`) for link, but with the incorrect suffix `.bc` that represent LLVM bitcode files.

> Independently Godot may want to change the build system to use `.a` file suffixes for static libraries.

Absolutely. Thank you for noticing! That was some 4 years old configuration that I overlooked.
We'll update our build system to use the proper `.a` suffix.

**Comment by RReverser** [comment]
I'm running into missing `ntohs` in 2.0.23 too and this issue seemed the most relevant. Luckily it happens only at `-O0` and goes away with optimisations in my examples (I guess the referencing code gets DCEd), but it would be still useful to fix.

I've tried applying https://github.com/emscripten-core/emscripten/pull/14053 to test it out but getting some other errors:

```
cache:INFO: generating system headers: sysroot_install.stamp... (this will be cached in "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/cache/sysroot_install.stamp" for subsequent builds)
cache:INFO:  - ok
Traceback (most recent call last):
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/emcc.py", line 3731, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/emcc.py", line 3724, in main
    ret = run(args)
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/emcc.py", line 1109, in run
    phase_calculate_system_libraries(state, linker_arguments, linker_inputs, newargs)
  File "/usr/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/emcc.py", line 2445, in phase_calculate_system_libraries
    extra_files_to_link += system_libs.calculate([f for _, f in sorted(linker_inputs)] + extra_files_to_link, forced=state.forced_stdlibs)
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/tools/system_libs.py", line 1461, in calculate
    handle_reverse_deps(input_files)
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/tools/system_libs.py", line 1432, in handle_reverse_deps
    symbolses = building.llvm_nm_multiple([os.path.abspath(t) for t in input_files])
  File "/usr/local/google/home/rreverser/emsdk/upstream/emscripten/tools/building.py", line 210, in llvm_nm_multiple
    for key, value in parse_llvm_nm_symbols(results.stdout, llvm_nm_files):
ValueError: too many values to unpack (expected 2)
```

**Comment by RReverser** [comment]
Looks like even with a fix that patch doesn't help, sadly.

```
error: undefined symbol: ntohs (referenced by $inetNtop6__deps: ['$inetNtop4','ntohs'], referenced by $readSockaddr__deps: ['$Sockets','$inetNtop4','$inetNtop6','ntohs'], referenced by $getSocketAddress__deps: ['$readSockaddr'], referenced by __sys_connect__deps: ['$getSocketFromFD','$getSocketAddress','$SYSCALLS'], referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _ntohs may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
```

**Comment by felipellrocha** [comment]
Was this fixed? I'm running into this at 2.0.32

**Comment by sbc100** [comment]
Thanks.  Fix is in https://github.com/emscripten-core/emscripten/pull/15304.
